python/ray/data/_internal/stats.py

```diff
@@ -339,61 +343,115 @@ def _check_cluster_stats_actor():
         _stats_actor_cluster_id = current_cluster_id
 
 
-def update_stats_actor_metrics(
-    op_metrics: List[OpRuntimeMetrics],
-    tags_list: List[Dict[str, str]],
-    state: Dict[str, Any],
-):
-    global _stats_actor
-    _check_cluster_stats_actor()
+class StatsActorManager:
+    """A Class containing util functions that manage remote calls to _StatsActor.
 
-    _stats_actor.update_metrics.remote(
-        [metric.as_dict() for metric in op_metrics], tags_list, state
-    )
+    Currently there are two codepaths that report stats to _StatsActor.
+        1. Execution:
+            Updates are called from the driver process in _scheduling_loop_step.
+            An update is passed through to _StatsActor via remote call if the last
+            update was more than `STATS_ACTOR_UPDATE_INTERVAL_SECONDS` seconds ago.
+        2. Iteration:
+            Updates are called from wherever iteration takes place in iter_batches.
+            This class runs a background thread that takes the last known iteration
+            stats and updates _StatsActor every `STATS_ACTOR_UPDATE_INTERVAL_SECONDS`
+            seconds.
 
+    """
 
-def update_stats_actor_iter_metrics(stats: "DatasetStats", tags_list: Dict[str, str]):
-    global _stats_actor
-    _check_cluster_stats_actor()
+    def __init__(self):
+        self._last_update = 0
+        self._last_iter_stats_args = None
+        self._iter_update_thread = None
+        self._iter_update_condition = threading.Condition()
+        self._collect_iter_stats = False
 
-    _stats_actor.update_iter_metrics.remote(stats, tags_list)
+    # Execution methods
 
+    def update_stats_actor_metrics(
+        self,
+        op_metrics: List[OpRuntimeMetrics],
+        tags_list: List[Dict[str, str]],
+        state: Dict[str, Any],
+        force_update: bool = False,
+    ):
+        if (
+            force_update
+            or time.time() - self._last_update >= STATS_ACTOR_UPDATE_INTERVAL_SECONDS
+        ):
+            global _stats_actor
+            _check_cluster_stats_actor()
+            _stats_actor.update_metrics.remote(
+                [metric.as_dict() for metric in op_metrics], tags_list, state
+            )
+            self._last_update = time.time()
 
-def clear_stats_actor_metrics(tags_list: List[Dict[str, str]]):
-    global _stats_actor
-    _check_cluster_stats_actor()
+    def clear_stats_actor_metrics(self, tags_list: List[Dict[str, str]]):
+        global _stats_actor
+        _check_cluster_stats_actor()
 
-    _stats_actor.clear_metrics.remote(tags_list)
+        _stats_actor.clear_metrics.remote(tags_list)
 
+    # Iteration methods
 
-def clear_stats_actor_iter_metrics(tags: Dict[str, str]):
-    global _stats_actor
-    _check_cluster_stats_actor()
+    def update_stats_actor_iter_metrics(
```

Comment:
Is it possible for this to be called by multiple iterators at once?

Reply:
In theory yes, but the common case is that iterators will be in different processes.

What can happen if they're in the same process?

Reply:
This isn't thread safe right now

Reply:
It's not a common practice, but theoretically possible. Can we pass in a tag to distinguish them? This tag will be useful for the dashboard as well. Basically, I think we need the dataset id + execution id + StreamSplitDataIterator index in this tag.

Reply:
I added support for multiple iterators: we will store the last call from each iterator, and the background thread will send all of them to the stats actor at once.

For multiple iterators of the same dataset on the same process, it should be uncommon for their iteration to overlap right? if their iteration is non-overlapping then it should be pretty easy to see in the time-series view the separate iterations. And even if we do use execution id, is there a way for the user to get the id for themselves?

For streaming split iterator, I added the index.